### PR TITLE
Group dep node data into a single structure

### PR DESCRIPTION
**Pull request by Zoxc**
r? @michaelwoerister

### CODE REVIEW

**Comment by Zoxc** [comment]
@rust-timer build 2f088a6151d06e930ecb997ba0859a9c32fc3b8a

**Comment by michaelwoerister** [comment]
Looks like pretty solid wins! I don't have time to review this in detail right now. One thing I noticed: It seems like the PR undoes https://github.com/rust-lang/rust/pull/49069 which @wesleywiser implemented a while ago per my request. It would be good to check if the wins here are due to the struct-of-arrays approach being a bad idea in this scenario. Maybe it's a bad trade-off, i.e. we trade faster serialization for bad cache locality during tracking.

File: src/librustc/hir/map/collector.rs [hunk 1]
```diff
@@ -139,9 +182,9 @@ impl<'a, 'hir> NodeCollector<'a, 'hir> {
         let node_hashes = self
             .hir_body_nodes
             .iter()
-            .fold(Fingerprint::ZERO, |fingerprint, &(def_path_hash, dep_node_index)| {
+            .fold(Fingerprint::ZERO, |fingerprint, &(def_path_hash, hash)| {
```
Comment (michaelwoerister) [inline src/librustc/hir/map/collector.rs]: Can you rename `fingerprint` to something like `combined_fingerprint` and `hash` to `fingerprint` here? That should make it easier to understand what's going on.

File: src/librustc/hir/map/collector.rs [hunk 2]
```diff
@@ -51,21 +51,69 @@ pub(super) struct NodeCollector<'a, 'hir> {
...
     // We are collecting DepNode::HirBody hashes here so we can compute the
     // crate hash from then later on.
-    hir_body_nodes: Vec<(DefPathHash, DepNodeIndex)>,
+    hir_body_nodes: Vec<(DefPathHash, Fingerprint)>,
+}
+
+fn input_dep_node_and_hash<'a, I>(
+    dep_graph: &DepGraph,
+    hcx: &mut StableHashingContext<'a>,
+    def_node: DepNode,
```
Comment (michaelwoerister) [inline src/librustc/hir/map/collector.rs]: This should be `dep_node` I guess?

File: src/librustc/dep_graph/graph.rs [hunk 3]
```diff
@@ -211,7 +199,7 @@ impl DepGraph {
                 reads: SmallVec::new(),
                 read_set: Default::default(),
             })),
-            |data, key, task| data.borrow_mut().complete_task(key, task))
+            |data, key, f, task| data.borrow_mut().complete_task(key, task, f))
```
Comment (michaelwoerister) [inline src/librustc/dep_graph/graph.rs]: Can you rename `f` to `fingerprint` here and at the other occurrences?

File: src/librustc/hir/map/collector.rs [hunk 4]
```diff
@@ -51,21 +51,69 @@ pub(super) struct NodeCollector<'a, 'hir> {
...
+    } else {
+        let mut stable_hasher = StableHasher::new();
+        input.hash_stable(hcx, &mut stable_hasher);
+        stable_hasher.finish()
+    };
+
+    (dep_node_index, hash)
+}
+
+fn hir_dep_nodes<'a, I>(
```
Comment (michaelwoerister) [inline src/librustc/hir/map/collector.rs]: Could you please rename this to `alloc_hir_dep_nodes`?

**Review by michaelwoerister — COMMENTED** [review]
Thanks again for the PR, @Zoxc! r=me with the nits addressed. 

I'm not quite happy with factoring of functions in the hir collector. It seems that it should be possible to make this a little less complicated. But I couldn't come up with something better either.

**Pull request merged**